src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -322,6 +324,299 @@ public virtual void Encode(TextWriter output, char[] value, int startIndex, int
             }
         }
 
+        public unsafe virtual OperationStatus Encode(ReadOnlySpan<char> source, Span<char> destination, out int charsConsumed, out int charsWritten, bool isFinalBlock = true)
+        {
+            // Optimization: Detect how much "doesn't require escaping" data exists at the beginning of the buffer,
+            // and memcpy it directly to the destination.
+
+            int idx = FindFirstCharacterToEncode(source);
+
+            int numCharsToCopy = idx;
+            if (numCharsToCopy < 0)
+            {
+                numCharsToCopy = source.Length;
+            }
+
+            if (!source.Slice(0, numCharsToCopy).TryCopyTo(destination))
+            {
+                // There wasn't enough room in the destination to copy over the entire source buffer.
+                // We'll instead copy over as much as we can and return Incomplete. We do need to
+                // account for the fact that we don't want to truncate a multi-char UTF-16 subsequence
+                // mid-sequence (since a subsequent slice and call to Encode would produce invalid
+                // data).
+
+                source = source.Slice(destination.Length);
```

Comment:
>destination [](start = 38, length = 11)

Is it possible someone can pass destination span with 0 length? that can make source.Length - 1 negative

Reply:
Shouldn't this be `source = source.Slice(0, destination.Length);`, if your objective is to only copy as much as what would fit into the destination?

Reply:
These issues are addressed in the UTF-8 implementation of the method. This particular line of code belongs to the UTF-16 implementation, which is being deleted since currently nobody calls it.